### PR TITLE
[ZkTracer] Move scripts to build and run ACT4 tests into `Makefile`

### LINKED ISSUE #3139: [ZkTracer] Move scripts to build and run ACT4 tests into `Makefile`
🔐 🔐 `Should this be a security advisory instead?` 🔐 🔐

## Description
Write a clear and concise description of what you want to happen. Add any considered drawbacks.

## Motivation
This section should indicate what value we are receiving from it.

## Tasks
- [ ] Medium-level task that needs to be implemented for this feature request to be considered complete.

## Acceptance criteria
- [ ] List tests that need to be implemented or checks to consider this task completed. 

## Risks
- [ ] These are risks that may prevent completion of the task as well as areas to pay special attention to

## Remember to
 - [ ] Add the `documentation` label in case there is an impact on the documentation
 - [ ] Add `priority` and `team` labels
 - [ ] Add Task for updating the Runbook or adding/updating existing metrics and alerts.

### PR DESCRIPTION
**Pull request by lorenzogentile404**
This PR implements issue(s) #

### Checklist

* [ ] I wrote new tests for my new core changes.
* [ ] I have successfully ran tests, style checker and build against my new changes locally.
* [ ] If this change is deployed to any environment (including Devnet), E2E test coverage exists or is included in this
  PR.
* [ ] I have informed the team of any breaking changes if there are any.

---

> [!NOTE]
> **Medium Risk**
> Moderate risk because it replaces the ACT4 driver scripts with new Makefile automation (including docker/host build modes) and adds a target that installs system-level dependencies and modifies shell rc PATH settings.
> 
> **Overview**
> Adds first-class Makefile support to **build and run ACT4 (riscv-arch-test) self-checking ELFs** and feed them through `zkc`, including new `build-act4`/`run-act4` targets that manage repo checkout/pinning, docker vs host builds, and per-test result/log artifacts under `act4/bin/`.
> 
> Introduces `make -C arithmetization install-act4-host-deps` to provision host toolchains (Homebrew, `mise`, RISC-V GCC tools, Sail, and macOS `z3` wiring) and updates docs accordingly. Removes the prior `act4/scripts/*.sh` driver scripts and updates the examples docs/cleanup targets (including changing the default `SP` value used in the memory layout docs/Makefile).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 1cbe09046a2124fc3a84f868a0ec310ca6329bb9. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: arithmetization/src/test/examples/Makefile [hunk 1]
```diff
@@ -201,3 +202,67 @@ blake-all:
...
+ACT4_DEBUG ?= false
+ACT4_ELF_DIR ?= $(ACT4_WORK_DIR)/linea-rv64im-zicclsm/elfs
+ACT4_LOGS ?= $(MAKEFILE_DIR)act4/bin/logs
+ACT4_RESULTS ?= $(MAKEFILE_DIR)act4/bin/results.txt
+
+# Build the ACT4 self-checking ELFs with the Linea ACT4 config.
+build-act4:
+	@set -u; \
+	verbosity=""; \
+	if [ "$(ACT4_DEBUG)" = "true" ]; then verbosity="DEBUG=true"; elif [ "$(ACT4_FAST)" = "true" ]; then verbosity="FAST=true"; fi; \
```
[the remark this answers is not in the document]
Comment (lorenzogentile404) [reply arithmetization/src/test/examples/Makefile]: they are supposed to not be capitalized.

File: arithmetization/src/test/examples/Makefile [hunk 2]
```diff
@@ -201,3 +216,56 @@ blake-all:
...
+	work_dir=$$(cd "$(ACT4_WORK_DIR)" && pwd); \
+	[ -d "$(ACT4_CONFIG_DIR)/linea-rv64im-zicclsm" ] || { echo "error: missing $(ACT4_CONFIG_DIR)/linea-rv64im-zicclsm"; exit 2; }; \
+	config_dir=$$(cd "$(ACT4_CONFIG_DIR)" && pwd); \
+	docker image inspect "$(ACT4_IMAGE)" >/dev/null 2>&1 || { echo "error: docker image '$(ACT4_IMAGE)' not found"; exit 2; }; \
+	echo "building ACT4 ELFs in $$work_dir"; \
+	docker run --rm \
+		-v "$$config_dir:/act4/config/cores/linea:ro" \
+		-v "$$work_dir:/act4/work" \
+		"$(ACT4_IMAGE)" \
+		bash -c "rm -rf /act4/work/linea-rv64im-zicclsm && CONFIG_FILES=config/cores/linea/linea-rv64im-zicclsm/test_config.yaml EXTENSIONS=$(ACT4_EXTENSIONS) $(if $(filter true,$(ACT4_DEBUG)),DEBUG=true) $(if $(filter true,$(ACT4_FAST)),FAST=true) make --jobs $(ACT4_JOBS) --keep-going"
```
[the remark this answers is not in the document]
Comment (lorenzogentile404) [reply arithmetization/src/test/examples/Makefile]: they do not need to be mutually exclusive https://github.com/riscv/riscv-arch-test/blob/act4/Makefile.

File: arithmetization/src/test/examples/Makefile [hunk 3]
```diff
@@ -201,3 +217,56 @@ blake-all:
...
+	rm -f "$(ACT4_LOGS)"/*; \
+	> "$(ACT4_RESULTS)"; \
+	command -v go >/dev/null 2>&1 || { echo "error: go not found"; exit 2; }; \
+	command -v zkc >/dev/null 2>&1 || { echo "error: zkc not found"; exit 2; }; \
+	count=0; pass=0; fail=0; \
+	for elf in $$(find "$(ACT4_ELF_DIR)" -name '*.elf' | sort); do \
+		name=$$(basename "$$elf" .elf); \
+		base="$(ACT4_LOGS)/$$name"; \
+		count=$$((count + 1)); \
+		go run $(GO_MAIN) "$$elf" "" "$(PROGRAM_OFFSET)" 0 "$(ENTRY_POINT)" > "$$base.json" 2> "$$base.json.err"; \
```
[the remark this answers is not in the document]
Comment (lorenzogentile404) [reply arithmetization/src/test/examples/Makefile:289]: act4 does not use `IN_BYTES`, as programs are self-contained, so 0 is fine.

File: arithmetization/src/test/examples/Makefile [hunk 4]
```diff
@@ -201,3 +217,56 @@ blake-all:
...
+	echo "building ACT4 ELFs in $$work_dir"; \
+	docker run --rm \
+		-v "$$config_dir:/act4/config/cores/linea:ro" \
+		-v "$$work_dir:/act4/work" \
+		"$(ACT4_IMAGE)" \
+		bash -c "rm -rf /act4/work/linea-rv64im-zicclsm && CONFIG_FILES=config/cores/linea/linea-rv64im-zicclsm/test_config.yaml EXTENSIONS=$(ACT4_EXTENSIONS) $(if $(filter true,$(ACT4_DEBUG)),DEBUG=true) $(if $(filter true,$(ACT4_FAST)),FAST=true) make --jobs $(ACT4_JOBS) --keep-going"
+
+# Convert each ACT4 ELF to JSON, run it through zkc, and print PASS/FAIL lines.
+# ACT4 writes messages and exits via ecall, so keep the line after each ecall marker.
+run-act4:
```
Comment (amkCha) [inline arithmetization/src/test/examples/Makefile]: can you make run-act4 depend on the build-act4 ? to avoid the error message ? Let's try to trim the script a bit

File: arithmetization/src/test/examples/Makefile [hunk 5]
```diff
@@ -201,3 +217,56 @@ blake-all:
...
+	@set -u; \
+	[ -d "$(ACT4_ELF_DIR)" ] || { echo "error: missing ELF dir '$(ACT4_ELF_DIR)'; run make build-act4 first"; exit 2; }; \
+	mkdir -p "$(ACT4_LOGS)"; \
+	rm -f "$(ACT4_LOGS)"/*; \
+	> "$(ACT4_RESULTS)"; \
+	command -v go >/dev/null 2>&1 || { echo "error: go not found"; exit 2; }; \
+	command -v zkc >/dev/null 2>&1 || { echo "error: zkc not found"; exit 2; }; \
```
Comment (amkCha) [inline arithmetization/src/test/examples/Makefile]: you could make it depend on `install-zkc`

File: arithmetization/src/test/examples/Makefile [hunk 6]
```diff
@@ -201,3 +217,56 @@ blake-all:
...
+# Build the ACT4 self-checking ELFs with the Linea ACT4 config
+build-act4:
+	@set -u; \
+	mkdir -p "$(ACT4_WORK_DIR)"; \
+	work_dir=$$(cd "$(ACT4_WORK_DIR)" && pwd); \
+	[ -d "$(ACT4_CONFIG_DIR)/linea-rv64im-zicclsm" ] || { echo "error: missing $(ACT4_CONFIG_DIR)/linea-rv64im-zicclsm"; exit 2; }; \
+	config_dir=$$(cd "$(ACT4_CONFIG_DIR)" && pwd); \
+	docker image inspect "$(ACT4_IMAGE)" >/dev/null 2>&1 || { echo "error: docker image '$(ACT4_IMAGE)' not found"; exit 2; }; \
+	echo "building ACT4 ELFs in $$work_dir"; \
+	docker run --rm \
```
Comment (amkCha) [inline arithmetization/src/test/examples/Makefile]: why do you run it in docker ? In the CI it spins a new instance so it's always clean - does it bring sth to the local launches maybe ?

File: arithmetization/src/test/examples/Makefile (hunk 4 above)
Comment (lorenzogentile404) [reply arithmetization/src/test/examples/Makefile]: yes. Good point.

File: arithmetization/src/test/examples/Makefile (hunk 5 above)
Comment (lorenzogentile404) [reply arithmetization/src/test/examples/Makefile]: `install-zkc` is in the `arithmetization` level Makefile. Shall I make it depend from that external targer? It sounds good to me.

File: arithmetization/src/test/examples/Makefile (hunk 6 above)
Comment (lorenzogentile404) [reply arithmetization/src/test/examples/Makefile]: It is just easier to have a ready to use environment to build the tests. Otherwise, configuration for MacOS should be differnt. I got stucked on the Sail step when I tried to configure my machine directly.

Comment (amkCha) [reply arithmetization/src/test/examples/Makefile]: what is Sail ?

File: arithmetization/src/test/examples/Makefile (hunk 5 above)
Comment (amkCha) [reply arithmetization/src/test/examples/Makefile]: good question, how should dependency work ? let's say we can inherit from a higher level Makefile ?

File: arithmetization/src/test/examples/README.md [hunk 7]
```diff
@@ -112,13 +118,66 @@ riscv-test compile <name>.<ext> VERIFY_ELF=true
...
+│               └── results.txt                 # PASS/FAIL summary
+└── riscv-arch-test/                            # ACT4 framework checkout
+```
+
+First of all, Docker container must be built:
+
+```bash
+git clone -b act4 https://github.com/riscv/riscv-arch-test.git
+cd riscv-arch-test
+docker build -t riscv-act4 .
```
Comment (amkCha) [inline arithmetization/src/test/examples/README.md]: shouldn't that be in the make build-act4 command line ?

File: arithmetization/src/test/examples/README.md [hunk 8]
```diff
@@ -112,13 +118,66 @@ riscv-test compile <name>.<ext> VERIFY_ELF=true
...
+```
+
+Then, from `linea-monorepo/arithmetization/src/test/examples`:
+
+```bash
+make build-act4
+make run-act4
```
Comment (amkCha) [inline arithmetization/src/test/examples/README.md]: you can remove the 1st step when the 2nd one will depend on it

File: arithmetization/src/test/examples/README.md [hunk 9]
```diff
@@ -112,13 +118,66 @@ riscv-test compile <name>.<ext> VERIFY_ELF=true
...
 ## Target ISA
 
 All programs are compiled targeting `RV64IM` accordingly to the [Ethereum zkVM standards](https://github.com/eth-act/zkvm-standards/blob/main/standards/riscv-target/target.md).
 Note that `Zicclsm` extension does not affect the generated ELF so it is omitted.
 Moreover, ABI being `LP64` (soft-float) is relevant only for float numbers, which we do not use, so it can be omitted as well.
 
+## ACT4
+
+The `Makefile` in this folder allows running tests from the [ACT4 framework](https://github.com/riscv/riscv-arch-test). It uses the configuration in `act4/config/linea-rv64im-zicclsm/`.
+The `riscv-arch-test` repository should have the same parent directory as `linea-monorepo`. The folder structure is the following:
```
Comment (amkCha) [inline arithmetization/src/test/examples/README.md]: can you add a description of the arch-test per say ? tests all opcodes, edge cases, used by the EF etc with a few references for the other teams to know that it's our baseline test

File: arithmetization/src/test/examples/Makefile (hunk 5 above)
Comment (lorenzogentile404) [reply arithmetization/src/test/examples/Makefile]: I believe it makes sense and we keep it only one way. The normal flow is doing the configuration first and then run the tests, so it feels natural.

File: arithmetization/src/test/examples/Makefile (hunk 6 above)
Comment (lorenzogentile404) [reply arithmetization/src/test/examples/Makefile]: It is a formal model of RISC-V the act4 framework depends on https://github.com/riscv/sail-riscv https://github.com/riscv/riscv-arch-test#4-risc-v-sail-reference-model .

File: arithmetization/src/test/examples/README.md (hunk 7 above)
Comment (lorenzogentile404) [reply arithmetization/src/test/examples/README.md]: yes, good point.

File: arithmetization/src/test/examples/README.md (hunk 8 above)
Comment (lorenzogentile404) [reply arithmetization/src/test/examples/README.md]: yes.

File: arithmetization/src/test/examples/README.md (hunk 9 above)
Comment (lorenzogentile404) [reply arithmetization/src/test/examples/README.md]: ok.

File: arithmetization/src/test/examples/Makefile (hunk 5 above)
Comment (lorenzogentile404) [reply arithmetization/src/test/examples/Makefile]: Actually, `go` and `zkc` are already assumed to be installed, as speciefied in the prerequisites in the README. I would just remove the check and keep it simple.

File: arithmetization/src/test/examples/Makefile (hunk 6 above)
Comment (amkCha) [reply arithmetization/src/test/examples/Makefile]: So on the repo, they say you install it with a curl `curl --location https://github.com/riscv/sail-riscv/releases/download/0.11/sail-riscv-$(uname)-$(arch).tar.gz | sudo tar xvz --directory=/path/to/install --strip-components=1` and that's it - what was the step that caused an issue ?

File: arithmetization/src/test/examples/Makefile [hunk 10]
```diff
@@ -66,6 +66,25 @@ ENTRY_POINT_FROM_ELF     = $(shell LANG=C riscv64-unknown-elf-readelf -h $(BIN)
...
+ACT4_REF        ?= 9798a554ce4139f472c9ccd3a18c9061d0f7024d
+ACT4_REPO       ?= $(abspath $(shell git -C $(MAKEFILE_DIR) rev-parse --show-toplevel)/../riscv-arch-test)
+ACT4_CONFIG     := $(ACT4_CONFIG_DIR)/linea-rv64im-zicclsm/test_config.yaml
+ACT4_IMAGE      := riscv-act4:$(ACT4_REF)
+ACT4_EXTENSIONS := I,M
+ACT4_JOBS       := 4
+ACT4_ELF_DIR    := $(ACT4_WORK_DIR)/linea-rv64im-zicclsm/elfs
+ACT4_LOGS       := $(MAKEFILE_DIR)act4/bin/logs
+ACT4_RESULTS    := $(MAKEFILE_DIR)act4/bin/results.txt
+ACT4_BUILD_MODE ?= docker
```
Comment (amkCha) [inline arithmetization/src/test/examples/Makefile]: can you put "host" by default

**Review by amkCha — DISMISSED** [review]

File: arithmetization/src/test/examples/Makefile [hunk 11]
```diff
@@ -201,3 +221,88 @@ blake-all:
...
+	@set -u; \
+	rm -f "$(ACT4_LOGS)"/*; \
+	mkdir -p "$(ACT4_LOGS)"; \
+	> "$(ACT4_RESULTS)"; \
+	count=0; pass=0; fail=0; \
+	for elf in $$(find "$(ACT4_ELF_DIR)" -name '*.elf' | sort); do \
+		name=$$(basename "$$elf" .elf); \
+		base="$(ACT4_LOGS)/$$name"; \
+		count=$$((count + 1)); \
+		go run $(GO_MAIN) "$$elf" "" "$(PROGRAM_OFFSET)" 0 "$(ENTRY_POINT)" > "$$base.json" 2> "$$base.json.err"; \
```
[the remark this answers is not in the document]
Comment (lorenzogentile404) [reply arithmetization/src/test/examples/Makefile:289]: ENTRY_POINT exists.

**Review by amkCha — APPROVED** [review]

**Comment by amkCha** [comment]
Nice job @lorenzogentile404 !
